### PR TITLE
[SD-224] Use dynamically-derived type configuration

### PR DESCRIPTION
**Pull request by georgeee**
Twin PR: https://github.com/serokell/snowdrop/pull/128

### CODE REVIEW

File: snowdrop-util/src/Snowdrop/Util/Lens.hs [hunk 1]
```diff
@@ -28,7 +28,7 @@ class HasGetter s a => HasLens s a where
 instance HasGetter (NonEmpty a) a where
     gett (x :| _) = x
 
-instance HasGetter a a where
+instance {-# OVERLAPPABLE #-} HasGetter a a where
```
Comment (pva701) [inline snowdrop-util/src/Snowdrop/Util/Lens.hs]: I remember the time when you were against `{-# OVERLAPPABLE #-}` syntax

File: snowdrop-core/src/Snowdrop/Core/ERoComp/Types.hs [hunk 2]
```diff
@@ -51,7 +51,7 @@ type family Ctx conf :: *
 --
 -- Type variables @id@, @value@ describe types of key and value of key-value storage.
 -- Variable @res@ denotes result of 'DbAccess' execution.
-data DbAccess (components :: [*]) (res :: *)
+data DbAccess conf (components :: [*]) (res :: *)
```
Comment (pva701) [inline snowdrop-core/src/Snowdrop/Core/ERoComp/Types.hs]: why this phantom type is needed here?
it's absolutely messy to populate all datatypes in framework by all available type-parameters

**Comment by georgeee** [comment]
> I disagree to add conf parameter to DbAccess, pls provide some solid comment and also drop comment in code

Because otherwise you need to explicitly pass `conf` parameter for every usage of `query`. And for every function that uses `query` and not any other method based on `DbAccessM`/`DbAccessU` like `modifyAccum`. And it breaks the possibility to derive type of `appConf`, it starts to require explicit passing of `conf` parameter, also the idea is for this type to be derived.

Adding phantom `conf` param fixes the issue.

And yeah, you need to explicitely pass `conf` param because you execute your action in `BaseM (BException conf) ...` and so `conf` needs to be specified, but is not "caught" by `DbAccess` data type. This leads to a warning which can be suppressed with `AllowAmbiguousTypes` but leads to problem with deriving type on top level, as I mentioned.

File: snowdrop-util/src/Snowdrop/Util/Lens.hs (hunk 1 above)
Comment (georgeee) [reply snowdrop-util/src/Snowdrop/Util/Lens.hs]: I'm still not a fan of it. This case I decided to apply this as quick fix to resolve the reason for the `{-# OVERLAPPABLE #-}` later

File: snowdrop-block/src/Snowdrop/Block/State.hs [hunk 3]
```diff
@@ -45,19 +45,38 @@ import           Snowdrop.Execution (DbComponents, ExpandRawTxsMode, ExpandableT
...
+instance
+  BlockHeader blkType1 ~ BlockHeader blkType2
+  => HasGetter (Union TxRaw ( BlockRawTxs blkType1 )) (Union TxRaw ( BlockRawTxs blkType2 )) where
+    gett =
+      union (union absurdUnion
+        (ulift . TxRaw @(CloseBlockRawTxType blkType2) . CloseBlockRawTx . unCloseBlockRawTx . unTxRaw) )
+      (ulift . TxRaw @(OpenBlockRawTxType blkType2) . OpenBlockRawTx . unOpenBlockRawTx . unTxRaw)
+
+instance
+  ( HasGetter (Union TxRaw (t2 : t3 : t4)) (Union TxRaw (t2' : t3' : t4'))
```
[the remark this answers is not in the document]
Comment (georgeee) [reply snowdrop-block/src/Snowdrop/Block/State.hs]: Oh, this is a hack and I need to think in future how to make it better

**Pull request merged**